[NL-6] : Splash 화면 구현

GomHyeok opened the pull request:
## 티켓 링크
https://nextershanbang.atlassian.net/browse/NL-6

## 구현내용
- SplashVC 화면 완성
- Splash 이벤트 관리 ViewModel 생성
- LaunchScreen 화면 생성
- LaunchScreen 노출 시간 설정

## 고민사항
- 화면 이동시 Lib Module의 Router 구현 방식을 어떻게 가져가는 것이 좋을지 의문입니다. 
- Combine 사용시 View와 ViewModel 사이의 이벤트 핸들링 방식과 ViewModel을 평소에 어떤 구조로 사용하시는지 말씀해주시면 맞추도록 하겠습니다!
- Combine 사용시 UIKit과 결합한 extension을 생성하여 사용하는 것에 대한 의견을 여쭤보고 싶습니다.

```swift
import Combine
import UIKit

extension UIControl {
    //주어진 이벤트에 대한 Publisher 반환(EventPublisher)
    func controlPublisher(for event: UIControl.Event) -> UIControl.EventPublisher {
        .init(control: self, event: event)
    }
    
    //UIControl객체와 이벤트 전달
    struct EventPublisher: Publisher {
        typealias Output = UIControl
        typealias Failure = Never
        
        let control: UIControl
        let event: UIControl.Event
        
        //Subscrier가 구독 시작시 호출
        func receive<S>(subscriber: S) where S : Subscriber, Never == S.Failure, UIControl == S.Input {
            let subscription = EventSubscription(control: control, subscrier: subscriber, event: event)
            subscriber.receive(subscription: subscription)
        }
    }
    
    //이벤트 발생할 때마다 구독자에게 이벤트 전달
    fileprivate class EventSubscription<EventSubscriber: Subscriber>: Subscription where EventSubscriber.Input == UIControl, EventSubscriber.Failure == Never {
        
        //control과 event를 통해 특정 이벤트를 수신할 수 있도록 설정
        let control: UIControl
        let event: UIControl.Event
        var subscriber: EventSubscriber?
        
        init(control: UIControl, subscrier: EventSubscriber, event: UIControl.Event) {
            self.control = control
            self.subscriber = subscrier
            self.event = event
            
            control.addTarget(self, action: #selector(handleEvent), for: event)
        }
        
        func request(_ demand: Subscribers.Demand) {}
        
        //구독을 취소할 때 호출
        func cancel() {
            subscriber = nil
            control.removeTarget(self, action: #selector(handleEvent), for: event)
        }
        
        //이벤트 발생 시 호출되는 함수 -> 구독자에게 UIControl 전달하여 이벤트 발생했음을 알림
        @objc func handleEvent() {
            _ = subscriber?.receive(control)
        }
    }
}

extension UIButton {
    var tapPublisher: AnyPublisher<Void, Never> {
        controlPublisher(for: .touchUpInside)
            .map { _ in }
            .eraseToAnyPublisher()
    }
}
```

## 기타

https://github.com/user-attachments/assets/a6b0c4d8-39a4-4c7b-a66a-13c4ee2c0a43

GomHyeok committed:
[NL-6] : SplashVC 생성
- SplashVC view 완성

GomHyeok committed:
[NL-6] : LaunchScreen 설정
- 로고 출력과 백그라운드 색상 변경
- appdelegate에서 screen 노출 시간 설정

GomHyeok committed:
[NL-6] : SplashViewModel 생성
- 시작하기 버튼 이벤트 발생시 화면 이동 할 수 있도록 setting

GomHyeok committed:
[NL-6] : Splash 오류 해결

TTOzzi commented on App/Sources/AppDelegate.swift:
```diff
@@ -6,15 +6,16 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
   func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
-    // Override point for customization after application launch.
+      
+    sleep(3)
```
메인 스레드를 재워 앱 실행 자체를 지연시키는 건 조금 위험할 것 같아요 🤔 
일단 자연스러운 흐름으로 실행되도록 구현하고, 나중에 런치 스크린을 더 길게 보여줘야 한다는 요구사항이 있다면 별도의 화면에 흐름을 만들어주는 게 어떨까요?

TTOzzi replied:
아 피그마를 방금 확인했는데 3초 노출이라는 조건이 있군요 🫠

GomHyeok replied:
3초 노출이라는 조건을 만족시키기 위해서 새로운 화면을 만드는 것이 좋을까요?

TTOzzi replied:
넵 메인 스레드를 블로킹하는 건 추후에 초기 설정과 관련된 코드들이 `didFinishLaunchingWithOptions` 시점에 위치하게 될텐데, 이러한 초기 설정 로직 전체를 지연시키는거라 런치 스크린과 동일한 UI 를 가진 화면을 만들어서 3초후에 이동되도록 하는 게 좋을 것 같아요.

GomHyeok replied:
넵~! 수정하겠습니다~!

TTOzzi commented on App/Sources/SceneDelegate.swift:
```diff
@@ -10,7 +11,8 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
     
-    let viewController = ViewController()
+    // TODO: Splash로 이동 or Home으로 이동 선택 로직 필요
+    let viewController = SplashViewcontroller()
```
화면 전환을 담당할 객체를 만들어주는 게 좋을 것 같긴 하네요.
같이 화면 구현하면서 천천히 고민해보시죠.

GomHyeok replied:
넵!

TTOzzi commented on Feature/Onboarding/Sources/Splash/SplashViewModel.swift:
```diff
@@ -0,0 +1,35 @@
+
+public protocol SplashViewModelProtocol : SplashViewModelInput, SplashViewModelOutput { }
+
+public final class SplashViewModel : SplashViewModelProtocol {
```
개인적으론 쉬운 디버깅을 위해 Input 을 enum 으로 묶어서 진입점을 한 곳으로 모아주는 것을 선호하긴 합니다만, 지금 구조도 좋습니다 👍

GomHyeok replied:
해당 구조로 변경할 수 있도록 하겠습니다~!

TTOzzi replied:
요건 강한 의견은 아니라서 반영하지 않아도 괜찮습니다!

TTOzzi commented on Feature/Onboarding/Sources/Splash/SplashViewcontroller.swift:
```diff
@@ -0,0 +1,75 @@
+
+public final class SplashViewcontroller : UIViewController {
+    private var store : [AnyCancellable] = []
+    private var viewModel: SplashViewModelProtocol!
```
`viewModel` 은 어떤 형태로 주입받게 되나요? 👀

GomHyeok replied:
setupBind함수를 통해 주입하거나 생성자 주입을 생각했습니다!

TTOzzi commented on Feature/Onboarding/Sources/Splash/SplashViewcontroller.swift:
```diff
@@ -0,0 +1,75 @@
+        $0.setAttributedTitle(styled, for: .normal)
+        $0.backgroundColor = DesignSystemAsset.Colors.primary2.color
+        $0.layer.cornerRadius = 8
+    }
```
프로퍼티, 생성자, 함수 선언 위치의 기준이 궁금해요 👀

GomHyeok replied:
view와 관련 없는 프로퍼티 -> 함수 -> view관련 프로퍼티 -> extension 순으로 작성하고 있었습니다! 다른 구조로 변경할까요?!?!

TTOzzi replied:
개인적인 생각으론 코드를 위에서 아래로 읽어내려갈 때, 지금 순서론 해당 프로퍼티에 접근하는 레이아웃 설정 함수들을 먼저 본 이후에 이 화면이 어떤 뷰들을 가지고 있는지를 확인할 수 있다보니 프로퍼티들을 위쪽으로 몰아주는 게 좋을 것 같아요 🤔

GomHyeok replied:
넵 수정하겠습니다!

TTOzzi reviewed (approved):
고생하셨습니다!

TTOzzi commented:
> Combine 사용시 UIKit과 결합한 extension을 생성하여 사용하는 것에 대한 의견을 여쭤보고 싶습니다.

말씀해주신 방법 좋습니다! 👍 
전달주신 코드에서 순환 참조 안 생기도록 `EventSubscription` 의 `control` 만 `weak` 로 선언해주면 좋을 것 같아요.

GomHyeok commented:
수정후 다시 PR 생성하도록 하겠습니다!

GomHyeok committed:
[NL-6] : Onbarding 화면 생성
- 성별 까지 구현 완료

# Conflicts:
#	App/Sources/AppDelegate.swift
#	Core/Project.swift

GomHyeok committed:
[NL-6] : TimeRange BottomSheet 생성

TTOzzi commented on App/Sources/AppDelegate.swift:
```diff
@@ -1,34 +1,37 @@
+        // Use this method to release any resources that were specific to the discarded scenes, as they will not return.
+    }
+    
+    private func dependencyInjection() {
```
요거 호출 안해줘도 되는 건가요? 👀

GomHyeok replied:
해당 부분은 router 부분이 모두 완성된 이후 한번에 수정하도록 하겠습니다!

TTOzzi commented on App/Sources/DIAseembly/CoreLayerAssembly.swift:
```diff
@@ -0,0 +1,21 @@
+
+import Swinject
+
+public class CoreLayerAssembly : Assembly {
+    public func assemble(container: Container) {
+        container.register(AppRouter.self) { _ in
+            return AppRouter.shared
+        }
+    }
+}
```
네이밍이 `CoreLayerAssembly` 인데, `AppRouter` 가 `CommonLayer` 의 `Lib` 에 있으니 `CommonLayerAssembly` 또는 `LibAssembly` 여야 하는거죠..? 
제가 잘 이해를 못한 것 같아 확인차 질문드립니다.

GomHyeok replied:
이름을 만드는 과정에서 제가 실수한 부분인 것 같습니다!

TTOzzi commented on Feature/Onboarding/Sources/Concreate/CheckBoxView.swift:
```diff
@@ -0,0 +1,83 @@
+
+    private let titleLabel = UILabel().then {
+        $0.text = "몰랐어요"
+        $0.font = Typography.Body_14_M.font?.font(size: 14) // 폰트 스타일 조정
```
typography 는 `style` 에 할당하고 `styledText` 에 텍스트 넣으심 됩니다!

GomHyeok replied:
넵! 해당 부분 수정하도록 하겠습니다!

TTOzzi commented on Feature/Onboarding/Sources/Concreate/DateTypeChipGroupView.swift:
```diff
@@ -0,0 +1,118 @@
+}
+
+// MARK: - DateTypeChipGroupView
+class DateTypeChipGroupView: UIView, DateTypeChipViewDelegate {
```
어떻게 생긴 컴포넌트인지 보려고 피그마 켰는데 음/양력 선택이 없어진 것 같네요 🤔

GomHyeok replied:
지금 확인했습니다! 원래 있었던 화면으로 기억하는데 일단 디자이너님께 여쭤봤습니다! 답변 결과에 따라 변경하도록 하겠습니다!

TTOzzi commented on Feature/Onboarding/Sources/Concreate/PickerButton.swift:
```diff
@@ -0,0 +1,134 @@
+
+class PickerButton : UIButton {
+    private let placeholderLabel = UILabel().then {
+        var style = Typography.Body_14_M
+        style.color = STColors.gray5.color
+        $0.attributedText = "Placeholder".set(style: style)
```
요렇게 설정하면 텍스트를 변경할 때마다 `.set(style: style)` 을 붙여주지 않으면 attributes 들이 초기화돼서 `.style` 에 할당해주는 게 편해요.

GomHyeok replied:
확인했습니다! 해당 부분들 다음 PR에서 모두 수정해도록 하겠습니다!

TTOzzi commented on Feature/Onboarding/Sources/Concreate/RadioButton.swift:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    // 라디오 버튼에 표시될 텍스트
+    var title: String? {
+        didSet {
+            let style = Typography.Body_14_M
+            style.color = DesignSystemAsset.Colors.black.color
+            titleLabel.attributedText = title?.set(style: style)
+        }
+    }
```
특정 뷰 컴포넌트와 동기화되어야 하는 값은 데이터의 원천을 둘로 나누는 것보다 getter, setter 로 titleLabel 의 text 와 엮어주는 게 좋을 것 같아요.
```suggestion
    var title: String? {
        get { titleLabel.text }
        set {
            let style = Typography.Body_14_M
            style.color = DesignSystemAsset.Colors.black.color
            titleLabel.attributedText = newValue?.set(style: style)
        }
    }
```

GomHyeok replied:
component 부분에서 해당 코드로 작성되 부분들이 있는 것 같아서 다른 부분도 확인 후 수정하겠습니다!

TTOzzi commented on Feature/Onboarding/Sources/Onboarding/OnboardingViewcontroller.swift:
```diff
@@ -0,0 +1,455 @@
+        $0.layer.masksToBounds = true
+        $0.leftView = UIView(frame: CGRect(x: 0, y: 0, width: 14, height: 0))
+        $0.leftViewMode = .always
+        $0.tag = 101
```
`tag` 는 사용하지 않는다면 제거해도 될 것 같아요.

GomHyeok replied:
넵 제거하도록 하겠습니다!

TTOzzi commented on Feature/Onboarding/Sources/Onboarding/OnboardingViewcontroller.swift:
```diff
@@ -0,0 +1,455 @@
+
+// MARK: functions
+extension OnboardingViewController {
+    func checkFormat(text : String) -> Bool {
```
이런 로직은 나중에 뷰모델을 구현하게 되면 옮겨주는 게 더 좋을 것 같아요.

GomHyeok replied:
onboarding 화면의 정상 기능 확인 후 연산이나 체크가 필요한 로직에 대하여 모두 viewmodel로 변경하도록 하겠습니다

TTOzzi commented on Feature/Onboarding/Sources/Onboarding/OnboardingViewcontroller.swift:
```diff
@@ -0,0 +1,455 @@
+        let nsCurrentText = currentText as NSString
+        let updatedText = currentText.replacingCharacters(in: stringRange, with: string)
+        
+        if textField == self.nameTextField {
```
사소하지만 인스턴스간 비교는 `===` 이 더 적절할 것 같아요.

GomHyeok replied:
넵 확인후 수정하도록 하겠습니다!

TTOzzi reviewed (approved):
고생하셨습니다! 👍 
PR 크기가 점점 커져서 우선 머지하고 넘어가시죠!
